bessctl/measurement_utils.py

```diff
@@ -0,0 +1,393 @@
+        """
+        Returns the dictionary of all RTT value if `percentile` is none.
+        Otherwise returns the `percentile`th percentile RTT.
+        """
+        if self.rtt is None:
+            return None
+        if percentile is None:
+            return self.rtt
+        return self.rtt[percentile]
+
+    def jitter(self, percentile=None):
+        """
+        Returns the dictionary of all jitter value if `percentile` is none.
+        Otherwise returns the `percentile`th percentile jitter.
+        """
+        if jitter is None:
+            return None
+        if percentile is None:
+            return self.jitter
+        return self.jitter[percentile]
+
+    def __str__(self):
+        fmt = '[in / out] pkts: {:.3f}M / {:.3f}M, drops: {:.3f}M / {:.3f}M, bytes: {} / {}'
+        rtt_data = []
+        if self.rtt is not None:
+            fmt += '\n'
+            fmt += ', '.join(['rtt_{}: {{:.3f}} us'.format(k)
+                             for k in sorted(self.rtt.keys())])
+            for k in sorted(self.rtt.keys()):
+                rtt_data.append(self.rtt[k] / 1e3)
+
+            fmt += '\n'
+            fmt += ', '.join(['jit_{}: {{:.3f}} us'.format(k)
+                             for k in sorted(self.jitter.keys())])
+            for k in sorted(self.jitter.keys()):
+                rtt_data.append(self.jitter[k] / 1e3)
+
+        return fmt.format(
+            self.pkts_in() / 1e6, self.pkts_out() / 1e6,
+            self.drops_in() / 1e6, self.drops_out() / 1e6,
+            self.bytes_in(), self.bytes_out(),
+            *rtt_data)
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class PortStatsGenerator(object):
+
+    def __init__(self, bess, tx_port, rx_port, measure=None, rtt_percentiles=list(), rate=False):
+        """
+        Creates a generator that produces PortStats. When `tx_port` and
+        `rx_port` are configured differently, the generated PortStats objects
+        will have outbound throughput reported as seen by `tx_port` and inbound
+        throughput will be reported as seen by `rx_port`.
+
+        If `rate` is true, generated port stats will be reported as rates
+        instead of cummulative values. If `rate` is true and RTT stats are being
+        collected, i.e. `measure` is not None, the generated PortStats objects
+        will have RTT stats reported as the average of those seen between
+        subsequent calls to `next()`.
+        """
+        self.bess = bess
+        self.tx_port = tx_port
+        self.rx_port = rx_port
+        self.measure = measure
+        self.rtt_percentiles = rtt_percentiles
+        self.rate = rate
+        self.old_throughput = None
+        self.old_rtt = None
+        self.old_jitter = None
+        self.last_check = time.time()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+    @staticmethod
+    def throughput_delta(old, new, secs):
+        delta = copy.copy(old)
```

Comment:
Consider using `old.copy()` here, and ... (continued next comment, can't comment on a big line range 😺 )

Reply:
Switched to `foo.copy()` where I could. Most of the copies here are happening on protobufs.